runtime/jvmti/heapify.cpp

```diff
@@ -0,0 +1,416 @@
+{
+	J9HashTable *objectMap = (J9HashTable*)walkState->userData1;
+	j9object_t obj = *objectSlot;
+	J9VMThread *stackObjectThread = walkState->walkThread;
+	if (objectIsStackAllocated(stackObjectThread, obj)) {
+		J9JVMTIObjectMap exemplar;
+		initializeExemplar(stackObjectThread, obj, &exemplar);
+		J9JVMTIObjectMap *entry = (J9JVMTIObjectMap*)hashTableFind(objectMap, &exemplar);
+		if (NULL != entry) {
+//			PORT_ACCESS_FROM_VMC(currentThread);
+//			j9tty_printf(PORTLIB, "<%p> Replacing %p with %p in slot %p\n", walkState->walkThread, obj, entry->heapObject, stackLocation);
+			*objectSlot = entry->heapObject;
+		}
+	}
+}
+
+/**
+ * Allocate the heap replacement for a stack-allocated object
+ *
+ * @param currentThread - the current thread
+ * @param object - the stack-allocated object
+ *
+ * @return New heap object or NULL on allocation failure
+ */
+static j9object_t
+allocateHeapObject(J9VMThread *currentThread, j9object_t object)
+{
+	j9object_t heapObject = NULL;
+	J9Class *objectClass = J9OBJECT_CLAZZ(currentThread, object);
+	UDATA classFlags = J9CLASS_FLAGS(objectClass);
+	J9MemoryManagerFunctions const * const mmFuncs = currentThread->javaVM->memoryManagerFunctions;
+#if 0
+	j9objectmonitor_t *lockEA = NULL;
+
+	if (LN_HAS_LOCKWORD(currentThread, object)) {
+		lockEA = J9OBJECT_MONITOR_EA(currentThread, object);
+	} else {
+		J9ObjectMonitor *objectMonitor = monitorTableAt(currentThread, object);
+		// Should probably just assert this is NULL
+		if (NULL != objectMonitor) {
+			lockEA = &(objectMonitor->alternateLockword);
+		}
+	}
+	if (NULL != lockEA) {
+		Assert_JVMTI_true(NULL == J9_LOAD_LOCKWORD(currentThread, lockEA));
+	}
+#endif
+	if (J9_ARE_ANY_BITS_SET(classFlags, J9AccClassArray)) {
+		U_32 size = J9INDEXABLEOBJECT_SIZE(currentThread, object);
+		heapObject = mmFuncs->J9AllocateIndexableObject(currentThread, objectClass, size, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+	} else {
+		heapObject = mmFuncs->J9AllocateObject(currentThread, objectClass, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+	}
+
+	return heapObject;
+}
+
+/**
+ * Allocate the heap replacements for all discovered stack-allocated objects
+ *
+ * @param currentThread - the current thread
+ * @param objectMap - the object map
+ *
+ * @return Array of heap objects or NULL on allocation failure
+ *
+ * The objects in the returned array are ordered by the object map iteration.
+ */
+static j9object_t
+allocateHeapObjects(J9VMThread *currentThread, J9HashTable *objectMap)
+{
+	uint32_t objectCount = hashTableGetCount(objectMap);
+	J9JavaVM *vm = currentThread->javaVM;
+	J9Class *jlObject = J9VMJAVALANGOBJECT_OR_NULL(vm);
+	J9Class *arrayClass = jlObject->arrayClass;
+	j9object_t objectArray = vm->memoryManagerFunctions->J9AllocateIndexableObject(currentThread, arrayClass, objectCount, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+	if (NULL != objectArray) {
+		uint32_t index = 0;
```

Comment:
please, try to unify the integer types throughout the file (either uint32_t style or U_32 style).

Reply:
The perils of copying code from multiple sources.

Reply:
I've moved to U_32 for now, though I think we should be moving to the C types in general.

Reply:
In GC world we've been gradually moving to C types, as we would touch files.